File: trunk/acce2/graphics/gui/textBox.py
from graphics.prims2d.rect import Rect
from graphics.prims2d.text import Text
from core.vec2 import Vec2

class TextBox(Rect):
    def onAttach(self):
        self.defaultVar("stroke",dict(color = (0.5,0.5,0.5)))
        self.defaultVar("highlightStroke",dict(color = (1,1,1)))
        self.defaultVar("fill",dict(color = (0,0,0)))
        self.defaultVar("text","")
        self.defaultVar("static",False)
        
        Rect.onAttach(self)
        
        self.edge = Rect(parent = self,
                         vars = dict(filled = False,
                                     material = self.stroke,
                                     static = False))
                                     
        self.textPrim = Text(parent = self,
                             vars = dict(text = self.text,
                                         material = self.stroke,
                                         static = False,
                                         position = Vec2(0.01,0.5),
                                         align = 'left'))
                                     
        self.active = False
        
    def msg_mouseButtonDown(self, args):
        if self in args['hitObjects']:
            self.active = True
            self.edge.material = self.highlightStroke
        else:
            self.active = False
            self.edge.material = self.stroke
            
    def addText(self, text):
        self.text += text
        self.textPrim.text += text
        
    def removeText(self, amount = 1):
        self.text = self.text[:-amount]
        self.textPrim.text = self.textPrim.text[:-amount]
        
    def clearText(self):
        self.text = ""
        self.textPrim.text = ""
            
    def msg_keyDown(self, args):
        if self.active:
            name = args['name']
            if name == 'backspace' and self.text:
                self.removeText()
            elif name == 'space':
                self.addText(' ')
            elif len(name) == 1:
                self.addText(name)

File: trunk/fhe5/Graphics/Gui/TextBox.cpp
#include "TextBox.h"
#include <fhe/math/Color.h>
#include <fhe/math/Vec2.h>

namespace fhe
{
    namespace Graphics
    {
        
        FHE_ASPECT(TextBox,SceneNode2);
        
        FHE_FUNC_IMPL(TextBox,on_attach)
        {
            VarMap defFocusStroke, defUnfocusStroke, defFocusFill, defUnfocusFill;

            defFocusStroke.setVar<Color>("color",Color(1,1,1,1));
            defUnfocusStroke.setVar<Color>("color",Color(0.5,0.5,0.5,1));
            defFocusFill.setVar<Color>("color",Color(0.5,0.5,0.5,1));
            defUnfocusFill.setVar<Color>("color",Color(0.25,0.25,0.25,1));

            getEntity()->defaultVar<VarMap>("focusFill",defFocusFill);
            getEntity()->defaultVar<VarMap>("unfocusFill",defUnfocusFill);
            getEntity()->defaultVar<VarMap>("focusStroke",defFocusStroke);
            getEntity()->defaultVar<VarMap>("unfocusStroke",defUnfocusStroke);

            m_bg = getEntity()->buildChild("bg");
            m_bg->buildAspect("Graphics/Prims/Rect");
            
            m_edge = m_bg->buildChild("edge");
            m_edge->setVar<bool>("filled",false);
            m_edge->buildAspect("Graphics/Prims/Rect");
            
            m_text = m_bg->buildChild("text");
            m_text->setVar<std::string>("align","center");
            m_text->setVar<Vec2>("pos",Vec2(0.5,0.5));
            m_text->buildAspect("Graphics/Prims/Text");
            
            updateFocus();
                
            return Var();
        }
        
        void TextBox::updateFocus()
        {
            bool focused = getEntity()->getVar<bool>("focused",false);
            if ( m_bg )
            {
                m_bg->setVar<VarMap>("material", focused ?
                    getEntity()->getVar<VarMap>("focusFill") : getEntity()->getVar<VarMap>("unfocusFill"));
            }
            if ( m_edge )
            {
                m_edge->setVar<VarMap>("material", focused ?
                    getEntity()->getVar<VarMap>("focusStroke") : getEntity()->getVar<VarMap>("unfocusStroke"));
            }
            if ( m_text )
            {
                m_text->setVar<VarMap>("material", focused ?
                    getEntity()->getVar<VarMap>("focusStroke") : getEntity()->getVar<VarMap>("unfocusStroke"));
            }
        }
        
        FHE_FUNC_IMPL(TextBox,msg_mouseButtonDown)
        {
            getEntity()->setVar<bool>("focused",m_bg->call("msg_mouseButtonDown",arg).get<bool>(false));
            updateFocus();
            return Var();
        }
        
        FHE_FUNC_IMPL(TextBox,msg_keyDown)
        {
            if ( getEntity()->getVar<bool>("focused",false) )
            {
                char key = arg.get<int>();
                
                if ( key == 8 )
                {
                    std::string text = m_text->getVar<std::string>("text","");
                    m_text->setVar<std::string>("text",text.substr(0,text.size()-1));
                }
                else if ( key == 13 )
                {
                    VarMap args;
                    args.setVar<std::string>("path",getEntity()->getPath());
                    args.setVar<std::string>("text",m_text->getVar<std::string>("text",""));
                    getEntity()->getRoot()->publish("textBoxAccepted",Var::build<VarMap>(args));
                }
                else if ( isprint(key) )
                {
                    m_text->setVar<std::string>("text",m_text->getVar<std::string>("text","") + key);
                }
            }
                
            return Var();
        }
        
    }
}
